index.js

```diff
@@ -35,30 +35,58 @@ module.exports = class SubscriptionDedupe {
   subscriptions: SubscriptionMap;
   */
 
-  subscribe(topic/*: string */) {
+  subscribe(topic /*: string */) {
     let existing = this.subscriptions[topic];
 
     if (!existing) {
       // Nothing found, create the subscription object
       existing = this.subscriptions[topic] = {
         promise: this.options.onSubscribe(topic),
-        refCount: 0
+        refCount: 0,
       };
+    } else if (existing && existing.refCount === 0) {
+      // This connection was closing, but did not finish. If it had,
+      // we wouldn't have found the subscription. 
+      existing.promise = existing.promise.then(() =>
+        this.options.onSubscribe(topic)
+      );
     }
 
     existing.refCount++;
     return existing.promise;
   }
 
-  unsubscribe(topic/*: string */) {
+  unsubscribe(topic /*: string */) {
     let existing = this.subscriptions[topic];
+    let promise;
 
     if (existing) {
-      existing.refCount--;
-      if (existing.refCount === 0) {
-        this.options.onUnsubscribe(topic);
-        delete this.subscriptions[topic];
+      if (existing.refCount > 1) {
+        existing.refCount--;
+      } else if (existing.refCount === 0) {
+        // Will happen if you unsubscribe too many times.
+        if (this.options.warnOnTooManyUnsubscribes !== false) {
+          console.warn(`Attempted to close dedupe subscription for topic "${topic}", but it was already closing.`);
+        }
+        return existing.promise;
+      } else {
+        // No more references. Close down the connection.
+        existing.refCount = 0;
+
+        // When reopening a connection, the first one will wait for the close before reopening.
+        // This way, multiple reopens will wait for each other.
+
+        // $FlowFixMe
+        promise = existing.promise = existing.promise
+          .then(() => this.options.onUnsubscribe(topic))
+          .then(() => {
+            if (existing.refCount <= 0) {
+              delete this.subscriptions[topic];
```

Comment:
Sequence in which this breaks (assuming all promises are pending):

1. Subscribe: Creates new subscription object. Stays pending.
2. Unsubscribe: Updates object, still waiting for (1).
3. Subscribe: Updates object, still waiting for (1).
4. Unsubscribe: Updates object, still waiting for (1).
5. Promises for (1) and (2) resolve. We remove the object in the `then` of (2). Promises for (3) and (4) are still pending.
6. Subscribe: Recreates subscription object without waiting for previous steps to complete.
7. Promises for (3) and (6) resolve.
8. Promise for (4) completes, removing the subscription. We now have an entry for a subscription we don't hold anymore.